sys/net/application_layer/gcoap/gcoap.c

```diff
@@ -119,23 +146,142 @@ static void *_event_loop(void *arg)
+                            SOCK_DTLS_1_2, SOCK_DTLS_SERVER) < 0) {
+            puts("gcoap: error creating DTLS sock");
+            sock_udp_close(&_sock_udp);
+            return 0;
+        }
+        dsm_init();
+        sock_dtls_event_init(&_sock_dtls, &_queue, _on_sock_dtls_evt,
+                            NULL);
+#endif
+    } else {
+        sock_udp_event_init(&_sock_udp, &_queue, _on_sock_udp_evt, NULL);
+    }
+
+    event_loop(&_queue);
     return 0;
 }
 
+#if IS_ACTIVE(CONFIG_GCOAP_ENABLE_DTLS)
+/* Handles DTLS socket events from the event queue */
+static void _on_sock_dtls_evt(sock_dtls_t *sock, sock_async_flags_t type, void *arg) {
+    (void)arg;
+    coap_socket_t socket = { .type = COAP_SOCKET_TYPE_DTLS, .socket.dtls = sock};
+
+    if (type & SOCK_ASYNC_CONN_RECV) {
+        ssize_t res = sock_dtls_recv(sock, &socket.ctx_dtls_session,
+                            _listen_buf, sizeof(_listen_buf),
+                            CONFIG_GCOAP_DTLS_HANDSHAKE_TIMEOUT_USEC);
+        if (res != -SOCK_DTLS_HANDSHAKE) {
+            DEBUG("gcoap: could not establish DTLS session: %d\n", res);
+            sock_dtls_session_destroy(sock, &socket.ctx_dtls_session);
+            return;
+        }
+        dsm_state_t prev_state = dsm_store(sock, &socket.ctx_dtls_session,
+                                           SESSION_STATE_ESTABLISHED, false);
+
+        /* If session is already stored and the state was SESSION_STATE_HANDSHAKE
+        before, the handshake has been initiated internally by a gcoap client request
+        and another thread is waiting for the handshake. Send message to the
+        waiting thread to inform about established session */
+        if (prev_state == SESSION_STATE_HANDSHAKE) {
+            msg_t msg = { .type = DTLS_EVENT_CONNECTED };
+            msg_send(&msg, _auth_waiting_thread);
+        } else if (prev_state == NO_SPACE) {
+            /* No space in session management. Should not happen. If it occurs,
+            we lost track of sessions stored in tinydtls internally */
+            sock_dtls_session_destroy(sock, &socket.ctx_dtls_session);
+        }
+
+        /* If not enough session slots left: set timeout to free up session */
+        uint8_t minimum_free = CONFIG_GCOAP_DTLS_MINIMUM_AVAILABLE_SESSIONS;
+        if (dsm_get_num_available_slots() < minimum_free)
+        {
+            uint32_t timeout = CONFIG_GCOAP_DTLS_MINIMUM_AVAILABLE_SESSIONS_TIMEOUT_USEC;
+            event_callback_init(&_dtls_session_free_up_tmout_cb,
+                                _dtls_free_up_session, NULL);
+            event_timeout_init(&_dtls_session_free_up_tmout, &_queue,
+                               &_dtls_session_free_up_tmout_cb.super);
+            event_timeout_set(&_dtls_session_free_up_tmout, timeout);
+        }
+    }
+
+    if (type & SOCK_ASYNC_CONN_FIN) {
+        if (sock_dtls_get_event_session(sock, &socket.ctx_dtls_session)) {
+            dsm_remove(sock, &socket.ctx_dtls_session);
```

Comment:
No `sock_dtls_session_destroy()` here?

Reply:
`SOCK_ASYNC_CONN_FIN` is called when a session is closed from the remote peer. In that case, the session is already in closed state from tinyDTLS at this point. Since no other implementation exists, no idea how they would behave here. Would expect the same.

Reply:
Maybe add a comment that the session is already destroyed - everywhere else `dsm_remove()` and `sock_dtls_session_destroy()` always appear together (so I was wondering if it is justified to move `sock_dtls_session_destroy()` into `dsm_remove()`)